### PR TITLE
add functions to OScanner for getting types and checking that types are correct

### PR DESCRIPTION
**Pull request by DeathEgg**
What it says on the tin. Does all of the checks that otherwise would have to be handwritten every time. Also will be useful for replacing sc_man.h/.cpp in the last places it's used in.

### CODE REVIEW

File: common/oscanner.cpp [hunk 1]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+
+	// remove comma if necessary
+	if (str[str.length() - 1] == ',')
+	{
+		str[str.length() - 1] = '\0';
+	}
+
+	if (IsNum(str.c_str()) == false || str != "MAXINT")
+	{
+		error("Missing integer (unexpected end of file).");
```
Comment (AlexMax) [inline common/oscanner.cpp:432]: This doesn't seem to me like an EOF condition

File: common/oscanner.cpp [hunk 2]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+
+	// remove comma if necessary
+	if (str[str.length() - 1] == ',')
+	{
+		str[str.length() - 1] = '\0';
+	}
+
+	if (IsRealNum(str.c_str()) == false)
+	{
+		error("Missing floating-point number (unexpected end of file).");
```
Comment (AlexMax) [inline common/oscanner.cpp:457]: This also does not look to me like an EOF condition.

File: common/oscanner.cpp [hunk 3]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+void OScanner::mustGetBool()
+{
+	if (!scan())
+	{
+		error("Missing boolean (unexpected end of file).");
+	}
+	
+	if (stricmp(_token.c_str(), "true") != 0 && stricmp(_token.c_str(), "false") != 0)
+	{
+		error("Missing boolean (unexpected end of file).");
```
Comment (AlexMax) [inline common/oscanner.cpp:473]: This also doesn't look like an yadda yadda you get the picture. 😃

**Review by leximayfield — APPROVED** [review]

**Comment by leximayfield** [comment]
@DeathEgg I accidently hit "approve" instead of "Request Changes".

File: common/oscanner.cpp [hunk 4]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+//
+// Get token as an int.
+//
+int OScanner::getTokenAsInt() const
+{
+	// fix for parser reading in commas
+	std::string str = _token;
+
+	// remove comma if necessary
+	if (str[str.length() - 1] == ',')
```
Comment (leximayfield) [inline common/oscanner.cpp]: Can't you use std::string::back() here?

File: common/oscanner.cpp (hunk 1 above)
Comment (leximayfield) [inline common/oscanner.cpp]: Seems like a bad copy-paste of an error message.

File: common/oscanner.cpp [hunk 5]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+	if (!scan())
+	{
+		error("Missing integer (unexpected end of file).");
+	}
+
+	// fix for parser reading in commas
+	std::string str = _token;
+
+	// remove comma if necessary
+	if (str[str.length() - 1] == ',')
```
Comment (leximayfield) [inline common/oscanner.cpp]: std::string::back()

File: common/oscanner.cpp [hunk 6]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+//
+// Get token as a float.
+//
+float OScanner::getTokenAsFloat() const
+{
+	// fix for parser reading in commas
+	std::string str = _token;
+
+	// remove comma if necessary
+	if (str[str.length() - 1] == ',')
```
Comment (leximayfield) [inline common/oscanner.cpp]: std::string::back()

File: common/oscanner.cpp [hunk 7]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+	if (!scan())
+	{
+		error("Missing floating-point number (unexpected end of file).");
+	}
+
+	// fix for parser reading in commas
+	std::string str = _token;
+
+	// remove comma if necessary
+	if (str[str.length() - 1] == ',')
```
Comment (leximayfield) [inline common/oscanner.cpp]: std::string::back()

File: common/oscanner.cpp [hunk 8]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+	{
+		str[str.length() - 1] = '\0';
```
Comment (leximayfield) [inline common/oscanner.cpp]: Wouldn't it be better to resize the string instead of forcing a null byte here?

File: common/oscanner.cpp (hunk 3 above)
Comment (leximayfield) [inline common/oscanner.cpp]: Error message copy paste

File: common/oscanner.cpp [hunk 9]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+// Ensure next token is a bool.
+//
+void OScanner::mustGetBool()
+{
+	if (!scan())
+	{
+		error("Missing boolean (unexpected end of file).");
+	}
+	
+	if (stricmp(_token.c_str(), "true") != 0 && stricmp(_token.c_str(), "false") != 0)
```
Comment (leximayfield) [inline common/oscanner.cpp]: We have a function specifically for dealing with std::string case-insensitive comparisons:

https://github.com/odamex/odamex/blob/stable/common/cmdlib.cpp#L240

File: common/oscanner.cpp (hunk 2 above)
Comment (leximayfield) [inline common/oscanner.cpp]: Looks like an accidental error message copypaste.

File: common/oscanner.cpp [hunk 10]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+		error("Missing integer (unexpected end of file).");
+	}
+
+	// fix for parser reading in commas
+	std::string str = _token;
+
+	// remove comma if necessary
+	if (str[str.length() - 1] == ',')
+	{
+		str[str.length() - 1] = '\0';
```
Comment (leximayfield) [inline common/oscanner.cpp]: Can we not resize this string?  Forcing a null byte seems incorrect to me.

File: common/oscanner.cpp [hunk 11]
```diff
@@ -326,4 +332,146 @@ bool OScanner::isQuotedString() const
...
+
+	return static_cast<float>(num);
+}
+
+//
+// Get token as a bool.
+//
+bool OScanner::getTokenAsBool() const
+{
+	return stricmp(_token.c_str(), "true") == 0;
```
Comment (leximayfield) [inline common/oscanner.cpp]: We have a function specifically for dealing with std::string case-insensitive comparisons:

https://github.com/odamex/odamex/blob/stable/common/cmdlib.cpp#L240

**Review by leximayfield — CHANGES_REQUESTED** [review]
Got a few changes, but nothing serious.

File: common/oscanner.cpp (hunk 4 above)
Comment (DeathEgg) [reply common/oscanner.cpp]: fixed

**Review by leximayfield — APPROVED** [review]

**Comment by leximayfield** [comment]
LGTM, thanks for the fixes

**Pull request merged**